<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Business-Case-Study---Staff-Planning" data-toc-modified-id="Business-Case-Study---Staff-Planning-1">Business Case Study - Staff Planning</a></span></li><li><span><a href="#Problem-Statement" data-toc-modified-id="Problem-Statement-2">Problem Statement</a></span><ul class="toc-item"><li><span><a href="#Importing-the-libraries-and-reading-the-dataset" data-toc-modified-id="Importing-the-libraries-and-reading-the-dataset-2.1">Importing the libraries and reading the dataset</a></span></li></ul></li><li><span><a href="#Question-1" data-toc-modified-id="Question-1-3">Question 1</a></span><ul class="toc-item"><li><span><a href="#Mathematical-formulation-/-Pyomo-components" data-toc-modified-id="Mathematical-formulation-/-Pyomo-components-3.1">Mathematical formulation / Pyomo components</a></span></li></ul></li><li><span><a href="#Question-2" data-toc-modified-id="Question-2-4">Question 2</a></span><ul class="toc-item"><li><span><a href="#Data-pre-processing" data-toc-modified-id="Data-pre-processing-4.1">Data pre-processing</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Expected-output:" data-toc-modified-id="Expected-output:-4.1.0.1">Expected output:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Question-3" data-toc-modified-id="Question-3-5">Question 3</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Worst-case-and-best-case-analysis-based-on-the-staffs'-availability." data-toc-modified-id="Worst-case-and-best-case-analysis-based-on-the-staffs'-availability.-5.0.0.1">Worst-case and best-case analysis based on the staffs' availability.</a></span></li><li><span><a href="#3.1-Worst-case-analysis" data-toc-modified-id="3.1-Worst-case-analysis-5.0.0.2">3.1 Worst case analysis</a></span></li><li><span><a href="#3.2-Best-case-analysis" data-toc-modified-id="3.2-Best-case-analysis-5.0.0.3">3.2 Best case analysis</a></span></li><li><span><a href="#Expected-output:" data-toc-modified-id="Expected-output:-5.0.0.4">Expected output:</a></span></li></ul></li><li><span><a href="#3.1-Worst-Case-Analysis" data-toc-modified-id="3.1-Worst-Case-Analysis-5.0.1">3.1 Worst Case Analysis</a></span><ul class="toc-item"><li><span><a href="#3.1.1-Optimal-number-of-staff-members" data-toc-modified-id="3.1.1-Optimal-number-of-staff-members-5.0.1.1">3.1.1 Optimal number of staff members</a></span></li><li><span><a href="#3.1.2-Percentage-of-outsourced-applications" data-toc-modified-id="3.1.2-Percentage-of-outsourced-applications-5.0.1.2">3.1.2 Percentage of outsourced applications</a></span></li><li><span><a href="#3.1.3-Average-cost-per-application" data-toc-modified-id="3.1.3-Average-cost-per-application-5.0.1.3">3.1.3 Average cost per application</a></span></li></ul></li><li><span><a href="#3.2--Best-Case-Analysis" data-toc-modified-id="3.2--Best-Case-Analysis-5.0.2">3.2  Best Case Analysis</a></span><ul class="toc-item"><li><span><a href="#3.2.1-Optimal-number-of-staff-members" data-toc-modified-id="3.2.1-Optimal-number-of-staff-members-5.0.2.1">3.2.1 Optimal number of staff members</a></span></li><li><span><a href="#3.2.2-Percentage-of-outsourced-applications" data-toc-modified-id="3.2.2-Percentage-of-outsourced-applications-5.0.2.2">3.2.2 Percentage of outsourced applications</a></span></li><li><span><a href="#3.2.3-Average-cost-per-application" data-toc-modified-id="3.2.3-Average-cost-per-application-5.0.2.3">3.2.3 Average cost per application</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Question-4" data-toc-modified-id="Question-4-6">Question 4</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Creating-Visualisations" data-toc-modified-id="Creating-Visualisations-6.0.0.1">Creating Visualisations</a></span></li></ul></li></ul></li></ul></li></ul></div>

# Business Case Study - Staff Planning

# Problem Statement

 The insurance approval process, i.e., underwriting is one of the important and time-consuming tasks in an insurance application processing. When you submit your insurance application, the underwriter of the company evaluates it based on the details that you provide using a rule-based or an ML model and decides whether or not to approve your application. An insurance company InsurePlus wants you to help them with finding the optimal number of staff that they need for their insurance application approval process for the calendar year 2021. In the industry, the number of staffs is considered as a continuous variable. This is also called a staff Full-Time Equivalent. For example, if a full-time employee (FTE =1) works for 50 hours a week, 10 hours corresponds to 0.2 FTEs.

- The company operates in three states A, B, and C.
- The company can either handle an application with the staff that they hire or outsource it to a vendor. (Assume that there is no capacity limitation to outsourcing.)
- If they hire staff, he/she can handle 40 insurance applications per month when he/she works 100% of the workdays. However, there are days that he/she will be unavailable to process applications due to training, off days, etc.
- A staff member’s availability (in percentage) to work on processing the insurance applications for each month is shown in the table given below. As mentioned before, with 100% availability, each member can handle 40 applications.

A special note of practical relevance: In the industry, staff availability is predicted using a time-motion study. But in this case, you have been given fixed numbers for each month. States A and B have a regulatory restriction that the outsourced insurance applications cannot be more than 30% and 40% of the total number of applications for each month, respectively. The objective is to optimise the total cost for the application approval process by distributing the right number of applications between the FTEs and the vendors while meeting the monthly demand for each state at the same time.



## Importing the libraries and reading the dataset

In [1]:
# Importing Libraries 

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
from pyomo.environ import *

In [2]:
from __future__ import division
from pyomo.opt import SolverFactory

In [3]:
# Reading the data from Excel workbook - Staffing.xlsx

InputData = 'Staffing+Data.xlsx'

#Read the data from Demand Data sheet

Demand = pd.read_excel(InputData,sheet_name='DemandData')

#Read the data from Staff Availability sheet

Availability = pd.read_excel (InputData,sheet_name='StaffAvailability')

#Read the data from Cost sheet

Cost = pd.read_excel (InputData,sheet_name='Cost')

#Max Applications Handled by FTE per month

ServiceRate = pd.read_excel( InputData, sheet_name ='ServiceRate')

In [4]:
#Printing the dataframe Demand

Demand.head()

State Month  Demand
0     A   Jan    5240
1     A   Feb    4878
2     A   Mar    5942
3     A   Apr    2297
4     A   May    1992

In [5]:
#Printing the dataframe Availability

Availability.head()

State Month    LB    UB  StaffAvPer
0     A   Jan  0.70  0.90        0.81
1     A   Feb  0.65  0.85        0.76
2     A   Mar  0.70  0.80        0.75
3     A   Apr  0.75  0.85        0.80
4     A   May  0.70  0.85        0.78

In [6]:
#Printing the dataframe Cost

Cost.head()

State Month  AnnualSalary  MonthlySalary  UnitOutSourceCost
0     A   Jan         60000         5000.0                180
1     A   Feb         60000         5000.0                180
2     A   Mar         60000         5000.0                180
3     A   Apr         60000         5000.0                180
4     A   May         60000         5000.0                180

In [7]:
#Printing the dataframe ServiceRate

ServiceRate

MgAppServedPerMonth
0                   40

In [8]:
# Consolidating the three dataframes into single dataframe for further analysis

data=pd.merge(Demand,Availability ,on=['State','Month'])

Staff=pd.merge(data,Cost, on = ['State','Month'])

Staff.head()

State Month  Demand    LB    UB  StaffAvPer  AnnualSalary  MonthlySalary  \
0     A   Jan    5240  0.70  0.90        0.81         60000         5000.0   
1     A   Feb    4878  0.65  0.85        0.76         60000         5000.0   
2     A   Mar    5942  0.70  0.80        0.75         60000         5000.0   
3     A   Apr    2297  0.75  0.85        0.80         60000         5000.0   
4     A   May    1992  0.70  0.85        0.78         60000         5000.0   

   UnitOutSourceCost  
0                180  
1                180  
2                180  
3                180  
4                180

In [9]:
#Checking the Shape of Final dataframe

Staff.shape

(36, 9)

In [10]:
#Checking if any null values present in the final dataframe

Staff.isnull().sum()

State                0
Month                0
Demand               0
LB                   0
UB                   0
StaffAvPer           0
AnnualSalary         0
MonthlySalary        0
UnitOutSourceCost    0
dtype: int64

# Question 1

The company wants to know the optimised staffing recommendations for the business case described. 
Write the mathematical model for the deterministic optimisation problem. Define and explain your decision variables, objective function and the constraint. (Hint: Use months of the year as the model timeline).



##  Mathematical formulation / Pyomo components

---

**Sets:** 

The indexes for the given problem are, <br>
- State, $s \in State$
- Month, $m \in Month$

---

**Parameters:** 

- $\text{Demand}_{s,m} \text{ - Demand of Insurance Applications}$ <br>


- $\text{FTE}_{s,m} \text{ - Applications processed by FTE}$ <br>


- $\text{Outsource}_{s,m} \text{ - Applications outsourced}$ <br>


- $\text{FTE_Monthly_Salary}_{s,m} \text{ - Monthly Salary of FTE}$ <br>


- $\text{UnitOutSourceCost}_{s,m} \text{ - Unit Cost for outsourcing 1 application}$ <br>


- $\text{Staff_Availability}_{s,m} \text{ - Average availability of FTE}$ <br>


- $\text{Staff_Availability_LB}_{s,m} \text{ - Minimum availability of FTE}$ <br>


- $\text{Staff_Availability_UB}_{s,m} \text{ - Maximum availability of FTE}$ <br>


- $\text{MaxAppFTE} \text{ - Number of insurance applications that can be processed by an FTE in a month when working with 100 percent availability}$ <br>
                                                                                                                                                                                                                   
---

**Decision variables:**
                                                                                                             
- Total number of  applications processed by FTE for each state and month combination <br>

 $\text{FTE}_{s,m} \,\,\,\,\,\,\, \text{where} \ s \in State, m \in Month $<br>
 
 
- Total number of  applications outsourced for each state and month combination <br>

 $\text{Outsource}_{s,m} \,\,\,\,\,\,\, \text{where} \ s \in State, m \in Month $<br>


---

**Objective Function:** <br>

To minimize the total cost of insurance approval process by distributing the right number of applications between FTE and outsourced.



\begin{align}
\textrm{min} (\sum \limits _{s,m} \text{FTE}_{s,m}* \text{FTE_Monthly_Salary}_{s,m} + \sum \limits _{s,m} \text{Outsource}_{s,m}* \text{UnitOutSourceCost}_{s,m})
\end{align}
where $s \in State$ and $m \in Month$

---

**Constraints:**


- **Demand Constraint** :- Demand of Insurance for each state and month combination


>- $\sum \limits _{s,m} \text{FTE}_{s,m}* \text{Staff_Availability}_{s,m}*\text{MaxAppFTE} = \text{Demand} \ \ \ \ \forall  {s \in State, m \in Month}$


- **Staff Availability** :- Staff Availability for best and worst case scenerios.


>- $ \text{Staff_Availability}_{s,m}  = \text{Staff_Availability_LB}_{s,m} \ \ \ \ \forall  {s \in State, m \in Month}$

>- $ \text{Staff_Availability}_{s,m}  < \text{Staff_Availability_UB}_{s,m} \ \ \ \ \forall  {s \in State, m \in Month}$



- **Outsourcing Regulations** :- Regulatory Regulations on States A & B for outsourcing applications.

>- $ \text{Outsource}_{s,m}  < (\text0.3*\text{Demand}_{s,m}) \,\,\,\,\,\,\, \text{where} \ s \in A , m \in Month $

>- $ \text{Outsource}_{s,m}  < (\text0.4*\text{Demand}_{s,m}) \,\,\,\,\,\,\, \text{where} \ s \in B , m \in Month $


# Question 2

Code the problem is Python and use any optimization package to solve it. Add comments to your code to explain each step. 

## Data pre-processing 

In [11]:
# Create the required Python data structures for indexes and parameters

#Extracting States into a list

State = Staff['State'].unique()

# Extracting Months into a list

Month = Staff['Month'].unique()

# No of applications processed when FTE is 100% available

MaxAppFTE = ServiceRate['MgAppServedPerMonth']

# Create a Dictionary object Demand with index: 'State' and 'Month' and Value as "Demand" column from the Staff Dataframe

Demand = Staff.set_index(['State','Month'])['Demand'].to_dict()

# Create a Dictionary object FTE_Monthly_Salary with index: 'State' and 'Month' and Value as "MonthlySalary" column 
#from the Staff Dataframe

FTE_Monthly_Salary = Staff.set_index(['State','Month'])['MonthlySalary'].to_dict()

# Create a Dictionary object UnitOutSourceCost with index: 'State' and 'Month' and Value as "UnitOutSourceCost" column 
#from the Staff Dataframe

UnitOutSourceCost = Staff.set_index(['State','Month'])['UnitOutSourceCost'].to_dict()

# Create a Dictionary object Staff_Availability with index: 'State' and 'Month' and Value as "StaffAvPer" column 
#from the Staff Dataframe

Staff_Availability = Staff.set_index(['State','Month'])['StaffAvPer'].to_dict()

# Create a Dictionary object Staff_Availability_LB with index: 'State' and 'Month' and Value as "LB" column 
#from the Staff Dataframe

Staff_Availability_LB = Staff.set_index(['State','Month'])['LB'].to_dict()

# Create a Dictionary object Staff_Availability_UB with index: 'State' and 'Month' and Value as "UB" column 
#from the Staff Dataframe

Staff_Availability_UB = Staff.set_index(['State','Month'])['UB'].to_dict()

#### Expected output:

Create a data frame containing the number of outsourced applications  and the number of FTEs for each state-month combination. You can choose to have extra columns like staff availability, demand etc. in your dataframe apart from the ones mentioned earlier. 

In [12]:
# Creating a model instance

model = ConcreteModel()

In [13]:
# Define Pyomo sets and Parameters

model.s = Set(initialize=State.tolist(),doc='States')

model.m = Set(initialize=Month.tolist(),doc='Months')

model.demand = Param(model.s,model.m,initialize = Demand,doc='Demand')

model.sa = Param(model.s,model.m,initialize = Staff_Availability,doc = 'Staff_Availability')

In [14]:
# Decision variables

model.FTE = Var(model.s,model.m,doc='No of FTE',domain = NonNegativeReals)

model.Outsource = Var(model.s,model.m,doc='No of Outsource App',domain = NonNegativeIntegers)

In [15]:
# Constraints

# Demand Constraint

def total_demand(M,s,m):
    return (M.FTE[s,m]*40*M.sa[s,m] + M.Outsource[s,m] == M.demand[s,m])
            
model.total_demand = Constraint(model.s,model.m, rule=total_demand)

# Outsourcing Regulation

model.outsource_demand = ConstraintList()
for m in model.m:
    for s in model.s:
        if s == 'A':
            model.outsource_demand.add(expr = model.Outsource[s,m] <= 0.3*model.demand[s,m])
        elif s == 'B':
            model.outsource_demand.add(expr = model.Outsource[s,m] <= 0.4*model.demand[s,m])

In [16]:
# Objective function

model.value = Objective(expr = sum(sum(model.FTE[s,m]*FTE_Monthly_Salary[s,m]
                                       + model.Outsource[s,m]*UnitOutSourceCost[s,m] for s in model.s) for m in model.m), 
                        sense= minimize)

In [17]:
# Invoking the solver

result = SolverFactory('glpk').solve(model)
result.write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 17962336.4487699
  Upper bound: 17962336.4487699
  Number of objectives: 1
  Number of constraints: 61
  Number of variables: 73
  Number of nonzeros: 97
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.06705069541931152
# ----------------------------------------------------------
#   Solution Information
# ------------

In [18]:
#Printing the output of the Model

model.pprint()

8 Set Declarations
    FTE_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B', 'Mar'), ('B', 'Apr'), ('B', 'May'), ('B', 'Jun'), ('B', 'Jul'), ('B', 'Aug'), ('B', 'Sep'), ('B', 'Oct'), ('B', 'Nov'), ('B', 'Dec'), ('C', 'Jan'), ('C', 'Feb'), ('C', 'Mar'), ('C', 'Apr'), ('C', 'May'), ('C', 'Jun'), ('C', 'Jul'), ('C', 'Aug'), ('C', 'Sep'), ('C', 'Oct'), ('C', 'Nov'), ('C', 'Dec')}
    Outsource_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B

In [19]:
# Print the value of the objective function

Optimized_Cost = round(model.value()/1000000,2)
Optimized_Cost

17.96

**`Checkpoint 1:`** Seems like the company has to spend around 17.9 m$ in total for the application approval process.

In [20]:
#Creating an empty list

Output_Optimised = []

# For every iteration, the code below will create a list containing 9 values - State, Month, Demand,No of FTE,
#No of Outsource,No of FTE Processed Applications,Percent FTE Applications,Percent Outsourced Applications,Estimated Cost,
#Average Cost per Application

for s in model.s:
    for m in model.m:
        
        
        Demand_Optimised = model.demand[s, m]
        No_of_FTE=model.FTE[s,m].value
        No_of_Outsource = model.Outsource[s,m].value
        No_of_FTE_Processed_App = model.FTE[s,m].value*40*model.sa[s,m]
        Percent_FTE_App = round((No_of_FTE/Demand_Optimised)*100,2)
        Percent_Outsourced_App = round((No_of_Outsource/Demand_Optimised)*100,2)
        Estimated_Cost = model.FTE[s,m].value*FTE_Monthly_Salary[s,m] + model.Outsource[s,m].value*UnitOutSourceCost[s,m]
        Cost_Per_App = round(Estimated_Cost/Demand_Optimised,2)
        
        Output_Optimised.append([s,m,Demand_Optimised,No_of_FTE,No_of_Outsource,No_of_FTE_Processed_App,Percent_FTE_App, 
                              Percent_Outsourced_App,Estimated_Cost,Cost_Per_App])
        
print(Output_Optimised)

[['A', 'Jan', 5240, 161.728395061728, 0.0, 5239.999999999987, 3.09, 0.0, 808641.97530864, 154.32], ['A', 'Feb', 4878, 160.460526315789, 0.0, 4877.999999999985, 3.29, 0.0, 802302.631578945, 164.47], ['A', 'Mar', 5942, 198.066666666667, 0.0, 5942.000000000011, 3.33, 0.0, 990333.333333335, 166.67], ['A', 'Apr', 2297, 71.78125, 0.0, 2297.0, 3.12, 0.0, 358906.25, 156.25], ['A', 'May', 1992, 63.8461538461538, 0.0, 1991.9999999999986, 3.21, 0.0, 319230.76923076896, 160.26], ['A', 'Jun', 2275, 77.9109589041096, 0.0, 2275.0000000000005, 3.42, 0.0, 389554.794520548, 171.23], ['A', 'Jul', 5334, 137.279411764706, 1600.0, 3734.000000000003, 2.57, 30.0, 974397.05882353, 182.68], ['A', 'Aug', 3371, 110.888157894737, 0.0, 3371.0000000000055, 3.29, 0.0, 554440.7894736851, 164.47], ['A', 'Sep', 3759, 116.018518518519, 0.0, 3759.000000000016, 3.09, 0.0, 580092.592592595, 154.32], ['A', 'Oct', 3529, 120.856164383562, 0.0, 3529.0000000000105, 3.42, 0.0, 604280.82191781, 171.23], ['A', 'Nov', 4284, 110.2573

In [21]:
# Creating dataframe for the results

Output_Optimised = pd.DataFrame(Output_Optimised, columns = ['State','Month','Demand','No of FTE','No of Outsource',
                                                       'No of FTE Processed Applications','Percent FTE Applications',
                                                       'Percent Outsourced Applications','Estimated Cost',
                                                       'Cost per Application'])

Output_Optimised[['Estimated Cost in mn $']] = Output_Optimised[['Estimated Cost']]/1000000

Output_Optimised['Demand'] = Output_Optimised['Demand'].astype(int)

Output_Optimised['No of Outsource'] = Output_Optimised['No of Outsource'].astype(int)

Output_Optimised['No of FTE Processed Applications'] = Output_Optimised['No of FTE Processed Applications'].astype(int)

Output_Optimised = Output_Optimised.round({"No of FTE":1, "No of Outsource":0, "Estimated Cost in mn $":1})

In [22]:
#Printing Final dataframe

Output_Optimised

State Month  Demand  No of FTE  No of Outsource  \
0      A   Jan    5240      161.7                0   
1      A   Feb    4878      160.5                0   
2      A   Mar    5942      198.1                0   
3      A   Apr    2297       71.8                0   
4      A   May    1992       63.8                0   
5      A   Jun    2275       77.9                0   
6      A   Jul    5334      137.3             1600   
7      A   Aug    3371      110.9                0   
8      A   Sep    3759      116.0                0   
9      A   Oct    3529      120.9                0   
10     A   Nov    4284      110.3             1285   
11     A   Dec    5183      139.6             1554   
12     B   Jan    4927      152.1                0   
13     B   Feb    2628       51.9             1051   
14     B   Mar    2974       59.5             1189   
15     B   Apr    2338       73.1                0   
16     B   May    4020      128.8                0   
17     B   Jun    3147       64.7             1258   
18     B   Jul    4271       94.2             1708   
19     B   Aug    2620       51.7             1048   
20     B   Sep    4517      139.4                0   
21     B   Oct    4155       85.4             1662   
22     B   Nov    3137       69.2             1254   
23     B   Dec    4227       97.6             1690   
24     C   Jan    1162       35.9                0   
25     C   Feb    1967       64.7                0   
26     C   Mar    1898       63.3                0   
27     C   Apr    2261       70.7                0   
28     C   May    2030       65.1                0   
29     C   Jun    1642       56.2                0   
30     C   Jul    2489        0.0             2489   
31     C   Aug    2496       82.1                0   
32     C   Sep     922       28.5                0   
33     C   Oct    2421       82.9                0   
34     C   Nov     963        0.0              963   
35     C   Dec    1998        0.0             1998   

    No of FTE Processed Applications  Percent FTE Applications  \
0                               5239                      3.09   
1                               4877                      3.29   
2                               5942                      3.33   
3                               2297                      3.12   
4                               1991                      3.21   
5                               2275                      3.42   
6                               3734                      2.57   
7                               3371                      3.29   
8                               3759                      3.09   
9                               3529                      3.42   
10                              2998                      2.57   
11                              3628                      2.69   
12                              4927                      3.09   
13                              1577                      1.97   
14                              1785                      2.00   
15                              2338                      3.12   
16                              4020                      3.21   
17                              1888                      2.06   
18                              2563                      2.21   
19                              1572                      1.97   
20                              4517                      3.09   
21                              2492                      2.05   
22                              1883                      2.21   
23                              2537                      2.31   
24                              1162                      3.09   
25                              1967                      3.29   
26                              1898                      3.33   
27                              2261                      3.12   
28                              2030                      3.21   
29                      

In [23]:
# Writing the results in to an Excel sheet

with pd.ExcelWriter('Output_Optimised_Q2.xlsx') as writer:
    Output_Optimised.to_excel(writer)

# Question 3

#### Worst-case and best-case analysis based on the staffs' availability.

Assuming that the distribution is the same across all the states,

#### 3.1 Worst case analysis 

- 3.1.1 What is the optimal number of staff members for the worst case? 

- 3.1.2 What is the percentage of outsourcing for the worst case? 

- 3.1.3 What is the average cost per application for the worst case?


#### 3.2 Best case analysis 

- 3.2.1 What is the optimal number of staff members for the best case? 

- 3.2.2 What is the percentage of outsourcing for the best case? 

- 3.2.3 What is the average cost per application for the best case?


#### Expected output:

For each of the subtasks (3.1 and 3.2) create a data frame containing the number of outsourced applications and the number of FTEs for each state-month combination. You can choose to have extra columns like staff availability, demand etc. in your dataframe apart from the ones mentioned earlier. Also, print the overall average percentage of outsourced applications and the overall average cost per application. 

### 3.1 Worst Case Analysis 

In [24]:
# Creating a model instance

model1 = ConcreteModel()

In [25]:
# Define Pyomo sets and Parameters


model1.s = Set(initialize=State.tolist(),doc='States')

model1.m = Set(initialize=Month.tolist(),doc='Months')

model1.demand = Param(model1.s,model1.m,initialize = Demand,doc='Demand')

model1.sa = Param(model1.s,model1.m,initialize = Staff_Availability_LB,doc = 'Staff_Availability_LB')

In [26]:
# Decision variables

model1.FTE = Var(model1.s,model1.m,doc='No of FTE',domain = NonNegativeReals)

model1.Outsource = Var(model1.s,model1.m,doc='No of Outsource App',domain = NonNegativeIntegers)

In [27]:
# Constraints

def total_demand(M,s,m):
    return (M.FTE[s,m]*40*M.sa[s,m] + M.Outsource[s,m] == M.demand[s,m])


model1.total_demand = Constraint(model1.s, model1.m, rule=total_demand)

# Outsourcing Regulation

model1.outsource_demand = ConstraintList()
for m in model1.m:
    for s in model1.s:
        if s == 'A':
            model1.outsource_demand.add(expr = model1.Outsource[s,m] <= 0.3*model1.demand[s,m])
        elif s == 'B':
            model1.outsource_demand.add(expr = model1.Outsource[s,m] <= 0.4*model1.demand[s,m])

In [28]:
# Objective function

model1.value = Objective(expr = sum(sum(model1.FTE[s,m]*FTE_Monthly_Salary[s,m]
                                        + model1.Outsource[s,m]*UnitOutSourceCost[s,m] for s in model1.s) for m in model1.m),
                         sense= minimize)

In [29]:
# Invoking the solver

result = SolverFactory('glpk').solve(model1)
result.write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 19599482.5167888
  Upper bound: 19599482.5167888
  Number of objectives: 1
  Number of constraints: 61
  Number of variables: 73
  Number of nonzeros: 97
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.05765247344970703
# ----------------------------------------------------------
#   Solution Information
# ------------

In [30]:
# Creating dataframe for the results

model1.pprint()

8 Set Declarations
    FTE_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B', 'Mar'), ('B', 'Apr'), ('B', 'May'), ('B', 'Jun'), ('B', 'Jul'), ('B', 'Aug'), ('B', 'Sep'), ('B', 'Oct'), ('B', 'Nov'), ('B', 'Dec'), ('C', 'Jan'), ('C', 'Feb'), ('C', 'Mar'), ('C', 'Apr'), ('C', 'May'), ('C', 'Jun'), ('C', 'Jul'), ('C', 'Aug'), ('C', 'Sep'), ('C', 'Oct'), ('C', 'Nov'), ('C', 'Dec')}
    Outsource_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B

In [31]:
# Print the value of the objective function

Worst_Cost = round(model1.value()/1000000,2)
Worst_Cost

19.6

**`Checkpoint 2:`** The company has to spend around 19.6 m$ in total for the application approval process if the staffs are working with the minimum availability.

In [32]:
#Creating an empty list

Output_Worst = []

# For every iteration, the code below will create a list containing 9 values - State, Month, Demand,No of FTE,
#No of Outsource,No of FTE Processed Applications,Percent FTE Applications,Percent Outsourced Applications,Estimated Cost,
#Average Cost per Application

for s in model1.s:
    for m in model1.m:
        
        
        Demand_Worst = model1.demand[s, m]
        No_of_FTE=model1.FTE[s,m].value
        No_of_Outsource = model1.Outsource[s,m].value
        No_of_FTE_Processed_App = model1.FTE[s,m].value*40*model1.sa[s,m]
        Percent_FTE_App = round((No_of_FTE/Demand_Worst)*100,2)
        Percent_Outsourced_App = round((No_of_Outsource/Demand_Worst)*100,2)
        Estimated_Cost = model1.FTE[s,m].value*FTE_Monthly_Salary[s,m] + model1.Outsource[s,m].value*UnitOutSourceCost[s,m]
        Cost_Per_App = round(Estimated_Cost/Demand_Worst,2)
        
        Output_Worst.append([s,m,Demand_Worst,No_of_FTE,No_of_Outsource,No_of_FTE_Processed_App,Percent_FTE_App, 
                              Percent_Outsourced_App,Estimated_Cost,Cost_Per_App])
        
print(Output_Worst)

[['A', 'Jan', 5240, 187.142857142857, 0.0, 5239.999999999995, 3.57, 0.0, 935714.285714285, 178.57], ['A', 'Feb', 4878, 131.346153846154, 1463.0, 3415.0000000000045, 2.69, 29.99, 920070.7692307701, 188.62], ['A', 'Mar', 5942, 212.214285714286, 0.0, 5942.000000000008, 3.57, 0.0, 1061071.42857143, 178.57], ['A', 'Apr', 2297, 76.5666666666667, 0.0, 2297.0000000000014, 3.33, 0.0, 382833.33333333355, 166.67], ['A', 'May', 1992, 71.1428571428571, 0.0, 1991.9999999999986, 3.57, 0.0, 355714.2857142855, 178.57], ['A', 'Jun', 2275, 61.2692307692308, 682.0, 1593.0000000000011, 2.69, 29.98, 429106.153846154, 188.62], ['A', 'Jul', 5334, 155.583333333333, 1600.0, 3733.999999999992, 2.92, 30.0, 1065916.666666665, 199.83], ['A', 'Aug', 3371, 90.7692307692308, 1011.0, 2360.000000000001, 2.69, 29.99, 635826.153846154, 188.62], ['A', 'Sep', 3759, 134.25, 0.0, 3758.9999999999995, 3.57, 0.0, 671250.0, 178.57], ['A', 'Oct', 3529, 95.0384615384615, 1058.0, 2470.999999999999, 2.69, 29.98, 665632.3076923075, 18

In [33]:
# Creating dataframe for the results

Output_Worst = pd.DataFrame(Output_Worst, columns = ['State','Month','Demand','No of FTE','No of Outsource',
                                                       'No of FTE Processed Applications','Percent FTE Applications',
                                                       'Percent Outsourced Applications','Estimated Cost',
                                                       'Cost per Application'])

Output_Worst[['Estimated Cost in mn $']] = Output_Worst[['Estimated Cost']]/1000000

Output_Worst['Demand'] = Output_Worst['Demand'].astype(int)

Output_Worst['No of Outsource'] = Output_Worst['No of Outsource'].astype(int)

Output_Worst['No of FTE Processed Applications'] = Output_Worst['No of FTE Processed Applications'].astype(int)

Output_Worst = Output_Worst.round({"No of FTE":1, "No of Outsource":0, "Estimated Cost in mn $":1})

In [34]:
#Printing Final dataframe

Output_Worst

State Month  Demand  No of FTE  No of Outsource  \
0      A   Jan    5240      187.1                0   
1      A   Feb    4878      131.3             1463   
2      A   Mar    5942      212.2                0   
3      A   Apr    2297       76.6                0   
4      A   May    1992       71.1                0   
5      A   Jun    2275       61.3              682   
6      A   Jul    5334      155.6             1600   
7      A   Aug    3371       90.8             1011   
8      A   Sep    3759      134.2                0   
9      A   Oct    3529       95.0             1058   
10     A   Nov    4284      125.0             1285   
11     A   Dec    5183      151.2             1554   
12     B   Jan    4927      105.6             1970   
13     B   Feb    2628       60.7             1051   
14     B   Mar    2974       63.8             1189   
15     B   Apr    2338       46.8              935   
16     B   May    4020       86.1             1608   
17     B   Jun    3147       72.7             1258   
18     B   Jul    4271      106.8             1708   
19     B   Aug    2620       60.5             1048   
20     B   Sep    4517       96.8             1806   
21     B   Oct    4155       95.9             1662   
22     B   Nov    3137       78.5             1254   
23     B   Dec    4227      105.7             1690   
24     C   Jan    1162       41.5                0   
25     C   Feb    1967        0.0             1967   
26     C   Mar    1898       67.8                0   
27     C   Apr    2261       75.4                0   
28     C   May    2030       72.5                0   
29     C   Jun    1642        0.0             1642   
30     C   Jul    2489        0.0             2489   
31     C   Aug    2496        0.0             2496   
32     C   Sep     922       32.9                0   
33     C   Oct    2421        0.0             2421   
34     C   Nov     963        0.0              963   
35     C   Dec    1998        0.0             1998   

    No of FTE Processed Applications  Percent FTE Applications  \
0                               5239                      3.57   
1                               3415                      2.69   
2                               5942                      3.57   
3                               2297                      3.33   
4                               1991                      3.57   
5                               1593                      2.69   
6                               3733                      2.92   
7                               2360                      2.69   
8                               3758                      3.57   
9                               2470                      2.69   
10                              2998                      2.92   
11                              3628                      2.92   
12                              2957                      2.14   
13                              1577                      2.31   
14                              1785                      2.14   
15                              1403                      2.00   
16                              2411                      2.14   
17                              1889                      2.31   
18                              2563                      2.50   
19                              1572                      2.31   
20                              2711                      2.14   
21                              2493                      2.31   
22                              1882                      2.50   
23                              2536                      2.50   
24                              1162                      3.57   
25                                 0                      0.00   
26                              1898                      3.57   
27                              2261                      3.33   
28                              2029                      3.57   
29                      

#### 3.1.1 Optimal number of staff members 

In [35]:
# write your code here

Optimal_Staff_Members = round(Output_Worst['No of FTE'].mean(),1)

print("Optimal number of staff members : ", Optimal_Staff_Members)

Optimal number of staff members :  76.7


#### 3.1.2 Percentage of outsourced applications 

In [36]:
# write your code here

Total_Outsourcing =  sum(Output_Worst['No of Outsource'])

Total_Demand =  sum(Output_Worst['Demand'])

Percent_Outsourced_App_Total = round((Total_Outsourcing/Total_Demand)*100,1)

print("Percentage of outsourced applications : ", Percent_Outsourced_App_Total)

Percentage of outsourced applications :  35.1


#### 3.1.3 Average cost per application

In [37]:
# write your code here

Total_Cost = sum(Output_Worst['Estimated Cost'])

Total_Demand =  sum(Output_Worst['Demand'])

Average_Cost_Per_App = round(Total_Cost/Total_Demand,2)

print("Average cost per application: ", Average_Cost_Per_App)

Average cost per application:  173.0


In [38]:
# Writing the results in to an Excel sheet

with pd.ExcelWriter('Output_Worst_Q3.xlsx') as writer:
    Output_Worst.to_excel(writer)

### 3.2  Best Case Analysis 

In [39]:
# Creating a model instance

model2 = ConcreteModel()

In [40]:
# Define Pyomo sets and Parameters

model2.s = Set(initialize=State.tolist(),doc='States')

model2.m = Set(initialize=Month.tolist(),doc='Months')

model2.demand = Param(model2.s,model2.m,initialize = Demand,doc='Demand')

model2.sa = Param(model2.s,model2.m,initialize = Staff_Availability_UB,doc = 'Staff_Availability_UB')

In [41]:
# Decision variables

model2.FTE = Var(model2.s,model2.m,doc='No of FTE',domain = NonNegativeReals)

model2.Outsource = Var(model2.s,model2.m,doc='No of Outsource App',domain = NonNegativeIntegers)

In [42]:
# Constraints

# Demand Constraint

def total_demand(M,s,m):
    return (M.FTE[s,m]*40*M.sa[s,m] + M.Outsource[s,m] == M.demand[s,m])


model2.total_demand = Constraint(model2.s, model2.m, rule=total_demand)

# Outsourcing Regulation

model2.outsource_demand = ConstraintList()
for m in model2.m:
    for s in model2.s:
        if s == 'A':
            model2.outsource_demand.add(expr = model2.Outsource[s,m] <= 0.3*model2.demand[s,m])
        elif s == 'B':
            model2.outsource_demand.add(expr = model2.Outsource[s,m] <= 0.4*model2.demand[s,m])

In [43]:
# Objective function

model2.value = Objective(expr = sum(sum(model2.FTE[s,m]*FTE_Monthly_Salary[s,m]
                                        + model2.Outsource[s,m]*UnitOutSourceCost[s,m] for s in model2.s) for m in model2.m),
                         sense= minimize)


In [44]:
# Invoking the solver

result = SolverFactory('glpk').solve(model2)
result.write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 16527535.6379357
  Upper bound: 16527535.6379357
  Number of objectives: 1
  Number of constraints: 61
  Number of variables: 73
  Number of nonzeros: 97
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.049669504165649414
# ----------------------------------------------------------
#   Solution Information
# -----------

In [45]:
#Printing the Model

model2.pprint()

8 Set Declarations
    FTE_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B', 'Mar'), ('B', 'Apr'), ('B', 'May'), ('B', 'Jun'), ('B', 'Jul'), ('B', 'Aug'), ('B', 'Sep'), ('B', 'Oct'), ('B', 'Nov'), ('B', 'Dec'), ('C', 'Jan'), ('C', 'Feb'), ('C', 'Mar'), ('C', 'Apr'), ('C', 'May'), ('C', 'Jun'), ('C', 'Jul'), ('C', 'Aug'), ('C', 'Sep'), ('C', 'Oct'), ('C', 'Nov'), ('C', 'Dec')}
    Outsource_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain : Size : Members
        None :     2 :    s*m :   36 : {('A', 'Jan'), ('A', 'Feb'), ('A', 'Mar'), ('A', 'Apr'), ('A', 'May'), ('A', 'Jun'), ('A', 'Jul'), ('A', 'Aug'), ('A', 'Sep'), ('A', 'Oct'), ('A', 'Nov'), ('A', 'Dec'), ('B', 'Jan'), ('B', 'Feb'), ('B

In [46]:
# Print the value of the objective function

Best_Cost = round(model2.value()/1000000,2)
Best_Cost

16.53

**`Checkpoint 3:`** The company has to spend around 16.5 m$ in total for the application approval process if the staffs are working with the maximum availability.

In [47]:
#Creating an empty list

Output_Best = []

# For every iteration, the code below will create a list containing 9 values - State, Month, Demand,No of FTE,
#No of Outsource,No of FTE Processed Applications,Percent FTE Applications,Percent Outsourced Applications,Estimated Cost,
#Average Cost per Application

for s in model2.s:
    for m in model2.m:
        
        
        Demand_Best = model2.demand[s, m]
        No_of_FTE=model2.FTE[s,m].value
        No_of_Outsource = model2.Outsource[s,m].value
        No_of_FTE_Processed_App = model2.FTE[s,m].value*40*model2.sa[s,m]
        Percent_FTE_App = round((No_of_FTE/Demand_Best)*100,2)
        Percent_Outsourced_App = round((No_of_Outsource/Demand_Best)*100,2)
        Estimated_Cost = model2.FTE[s,m].value*FTE_Monthly_Salary[s,m] + model2.Outsource[s,m].value*UnitOutSourceCost[s,m]
        Cost_Per_App = round(Estimated_Cost/Demand_Best,2)
        
        Output_Best.append([s,m,Demand_Best,No_of_FTE,No_of_Outsource,No_of_FTE_Processed_App,Percent_FTE_App, 
                              Percent_Outsourced_App,Estimated_Cost,Cost_Per_App])
        
print(Output_Best)

[['A', 'Jan', 5240, 145.555555555556, 0.0, 5240.000000000016, 2.78, 0.0, 727777.77777778, 138.89], ['A', 'Feb', 4878, 143.470588235294, 0.0, 4877.999999999996, 2.94, 0.0, 717352.94117647, 147.06], ['A', 'Mar', 5942, 185.6875, 0.0, 5942.0, 3.12, 0.0, 928437.5, 156.25], ['A', 'Apr', 2297, 67.5588235294118, 0.0, 2297.0000000000014, 2.94, 0.0, 337794.11764705897, 147.06], ['A', 'May', 1992, 58.5882352941176, 0.0, 1991.9999999999984, 2.94, 0.0, 292941.176470588, 147.06], ['A', 'Jun', 2275, 71.09375, 0.0, 2275.0, 3.12, 0.0, 355468.75, 156.25], ['A', 'Jul', 5334, 177.8, 0.0, 5334.0, 3.33, 0.0, 889000.0, 166.67], ['A', 'Aug', 3371, 99.1470588235294, 0.0, 3370.9999999999995, 2.94, 0.0, 495735.29411764705, 147.06], ['A', 'Sep', 3759, 104.416666666667, 0.0, 3759.000000000012, 2.78, 0.0, 522083.333333335, 138.89], ['A', 'Oct', 3529, 110.28125, 0.0, 3529.0, 3.12, 0.0, 551406.25, 156.25], ['A', 'Nov', 4284, 142.8, 0.0, 4284.0, 3.33, 0.0, 714000.0, 166.67], ['A', 'Dec', 5183, 185.107142857143, 0.0, 5

In [48]:
# Creating dataframe for the results

Output_Best = pd.DataFrame(Output_Best, columns = ['State','Month','Demand','No of FTE','No of Outsource',
                                                       'No of FTE Processed Applications','Percent FTE Applications',
                                                       'Percent Outsourced Applications','Estimated Cost',
                                                       'Cost per Application'])

Output_Best[['Estimated Cost in mn $']] = Output_Best[['Estimated Cost']]/1000000

Output_Best['Demand'] = Output_Best['Demand'].astype(int)

Output_Best['No of Outsource'] = Output_Best['No of Outsource'].astype(int)

Output_Best['No of FTE Processed Applications'] = Output_Best['No of FTE Processed Applications'].astype(int)

Output_Best = Output_Best.round({"No of FTE":1, "No of Outsource":0, "Estimated Cost in mn $":1})

In [49]:
#Printing Final dataframe

Output_Best

State Month  Demand  No of FTE  No of Outsource  \
0      A   Jan    5240      145.6                0   
1      A   Feb    4878      143.5                0   
2      A   Mar    5942      185.7                0   
3      A   Apr    2297       67.6                0   
4      A   May    1992       58.6                0   
5      A   Jun    2275       71.1                0   
6      A   Jul    5334      177.8                0   
7      A   Aug    3371       99.1                0   
8      A   Sep    3759      104.4                0   
9      A   Oct    3529      110.3                0   
10     A   Nov    4284      142.8                0   
11     A   Dec    5183      185.1                0   
12     B   Jan    4927      136.9                0   
13     B   Feb    2628       77.3                0   
14     B   Mar    2974       92.9                0   
15     B   Apr    2338       68.8                0   
16     B   May    4020      118.2                0   
17     B   Jun    3147       98.3                0   
18     B   Jul    4271       85.4             1708   
19     B   Aug    2620       77.1                0   
20     B   Sep    4517      125.5                0   
21     B   Oct    4155      129.8                0   
22     B   Nov    3137       62.8             1254   
23     B   Dec    4227       90.6             1690   
24     C   Jan    1162       32.3                0   
25     C   Feb    1967       57.9                0   
26     C   Mar    1898       59.3                0   
27     C   Apr    2261       66.5                0   
28     C   May    2030       59.7                0   
29     C   Jun    1642       51.3                0   
30     C   Jul    2489       83.0                0   
31     C   Aug    2496       73.4                0   
32     C   Sep     922       25.6                0   
33     C   Oct    2421       75.7                0   
34     C   Nov     963       32.1                0   
35     C   Dec    1998       71.4                0   

    No of FTE Processed Applications  Percent FTE Applications  \
0                               5240                      2.78   
1                               4877                      2.94   
2                               5942                      3.12   
3                               2297                      2.94   
4                               1991                      2.94   
5                               2275                      3.12   
6                               5334                      3.33   
7                               3370                      2.94   
8                               3759                      2.78   
9                               3529                      3.12   
10                              4284                      3.33   
11                              5183                      3.57   
12                              4926                      2.78   
13                              2627                      2.94   
14                              2974                      3.12   
15                              2337                      2.94   
16                              4019                      2.94   
17                              3147                      3.12   
18                              2562                      2.00   
19                              2620                      2.94   
20                              4516                      2.78   
21                              4155                      3.12   
22                              1883                      2.00   
23                              2537                      2.14   
24                              1162                      2.78   
25                              1967                      2.94   
26                              1898                      3.12   
27                              2261                      2.94   
28                              2030                      2.94   
29                      

#### 3.2.1 Optimal number of staff members

In [50]:
# write your code here

Optimal_Staff_Members = round(Output_Best['No of FTE'].mean(),1)

print("Optimal number of staff members : ", Optimal_Staff_Members)

Optimal number of staff members :  92.9


#### 3.2.2 Percentage of outsourced applications

In [51]:
# write your code here

Total_Outsourcing =  sum(Output_Best['No of Outsource'])

Total_Demand =  sum(Output_Best['Demand'])

Percent_Outsourced_App_Total = round((Total_Outsourcing/Total_Demand)*100,1)

print("Percentage of outsourced applications : ", Percent_Outsourced_App_Total)

Percentage of outsourced applications :  4.1


#### 3.2.3 Average cost per application

In [52]:
# write your code here

Total_Cost = sum(Output_Best['Estimated Cost'])

Total_Demand =  sum(Output_Best['Demand'])

Average_Cost_Per_App = round(Total_Cost/Total_Demand,2)

print("Average cost per application: ", Average_Cost_Per_App)

Average cost per application:  145.88


In [53]:
# Writing the results in to an Excel sheet

with pd.ExcelWriter('Output_Best_Q3.xlsx') as writer:
    Output_Best.to_excel(writer)

# Question 4

#### Creating Visualisations

Create the following visualisations using your preferred method (i.e. Python, PowerPoint, Power BI, etc.) and add it to your report. 

Use the solution of Q2 to create a stacked column chart that shows the percentage of applications processed by the staff and by the vendor for each month (%staff processed applications+ %vendor processed applications should add up to 100%). 
Create a graph to show how the cost per application increases with respect to any change in the parameters in your analysis.
Hint: Use the cost per application that you calculate in Questions 2 and 3 (i.e., the best case, and the worst case). 

In [54]:
# We have attached the charts directly in Report

**Note:** You can create the charts in Python or some other visualisation tools and make it a part of your final report directly.